contribs/src/test/scala/io/delta/storage/ExternalLogStoreSuite.scala

```diff
@@ -0,0 +1,198 @@
+      assert(MemoryLogStore.hashMap.containsKey(path))
+    }
+  }
+
+  test("recovery") {
+    withSQLConf(
+      "fs.failing.impl" -> classOf[FailingFileSystem].getName,
+      "fs.failing.impl.disable.cache" -> "true"
+    ) {
+      FailingFileSystem.failOnSuffix = Some("00000000000000000000.json")
+      withTempDir { tempDir =>
+        val store = createLogStore(spark)
+        val path = FileNames.deltaFile(
+          new Path(new Path(s"failing:${tempDir.getCanonicalPath}"), "_delta_log"),
+          0
+        )
+        store.write(path, Iterator("foo", "bar"), false)
+        val entry = MemoryLogStore.hashMap.get(path)
+        assert(entry != null)
+        assert(!entry.complete)
+        assert(entry.tempPath.nonEmpty)
+
+        assertThrows[java.io.FileNotFoundException] {
+          store.read(path)
+        }
+        val contents = store.read(entry.absoluteTempPath()).toList
+
+        FailingFileSystem.failOnSuffix = None
+        val files = store.listFrom(path.toString)
+        val entry2 = MemoryLogStore.hashMap.get(path)
+        assert(entry2 != null)
+        assert(entry2.complete)
+        assert(store.read(entry2.absoluteJsonPath()).toList == contents)
+      }
+    }
+  }
+
+  test("listFrom exceptions") {
+    val store = createLogStore(spark)
+    assertThrows[java.io.FileNotFoundException] {
+      store.listFrom("/non-existing-path/with-parent")
+    }
+  }
+
+  protected def shouldUseRenameToWriteCheckpoint: Boolean = false
+}
+
+class MemoryLogStore(sparkConf: SparkConf, hadoopConf: Configuration)
+    extends BaseExternalLogStore(sparkConf, hadoopConf) {
+
+  import MemoryLogStore._;
+
+  override def invalidateCache(): Unit = {
+    hashMap.clear()
+  }
+
+  override protected def getLatestDbEntry(tablePath: Path): Option[LogEntry] = {
+    hashMap.asScala.iterator
+      .filter(r => r._2.tablePath == tablePath)
+      .toList
+      .sortBy(r => r._1)
+      .reverse
+      .find(r => true)
+      .map(r => r._2)
+  }
+
+  override protected def getDbEntry(tablePath: Path, path: Path): Option[LogEntry] = {
+    hashMap.asScala.get(path)
+  }
+
+  override protected def putDbEntry(
+      logEntry: LogEntry,
+      overwrite: Boolean = false
+  ): Unit = {
+
+    logDebug(
+      s"WriteExternalCache: ${logEntry.absoluteJsonPath} (overwrite=$overwrite)"
+    )
+
+    if (overwrite) {
+      hashMap.put(logEntry.absoluteJsonPath, logEntry)
+    } else {
+      val curr_val = hashMap.putIfAbsent(logEntry.absoluteJsonPath, logEntry);
+      if (curr_val != null) {
+        throw new java.nio.file.FileAlreadyExistsException(
+          s"TransactionLog exists ${logEntry.absoluteJsonPath}"
+        )
+      }
+    }
+  }
+}
+
+object MemoryLogStore {
+  val hashMap = new ConcurrentHashMap[Path, LogEntry]()
+}
+
+class TestLogStore(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration
+) extends MemoryLogStore(sparkConf, hadoopConf) {}
+
+class FailingFileSystem extends RawLocalFileSystem {
```

Comment:
can you add a bit of documentation here. I understand what the close does and how it is used, but can you add one or two lines explaining what it does + how + why we are using it. just short explanation is fine.